src/parent/props.js

```diff
@@ -1,103 +1,90 @@
-            continue;
+            // $FlowFixMe
+            newProps[key] = value;
+            return;
         }
 
         const alias = propDef.alias;
-        if (alias) {
-            if (!isDefined(value) && isDefined(inputProps[alias])) {
-                value = inputProps[alias];
-            }
-            aliases.push(alias);
+        if (alias && !isDefined(value) && isDefined(inputProps[alias])) {
+            value = inputProps[alias];
         }
 
         if (propDef.value) {
-            value = propDef.value({ props, state, close, focus, event, onError });
+            value = propDef.value({ props: newProps, state, close, focus, event, onError, container });
         }
 
-        if (!isDefined(value) && propDef.default) {
-            value = propDef.default({ props, state, close, focus, event, onError });
+        if (propDef.default && !isDefined(value) && !isDefined(newProps[key])) {
+            value = propDef.default({ props: newProps, state, close, focus, event, onError, container });
         }
 
         if (isDefined(value)) {
-            if (propDef.type === 'array' ? !Array.isArray(value) : (typeof value !== propDef.type)) {
-                throw new TypeError(`Prop is not of type ${ propDef.type }: ${ key }`);
-            }
+            // $FlowFixMe
+            newProps[key] = value;
         }
-        
-        // $FlowFixMe
-        props[key] = value;
-    }
-
-    for (const alias of aliases) {
-        delete props[alias];
-    }
+    });
 
-    eachProp(props, propsDef, (key, propDef, value) => {
+    eachProp(newProps, propsDef, (key, propDef, value) => {
         if (!propDef) {
             return;
         }
 
+        if (isDefined(value)) {
+            if (propDef.type === PROP_TYPE.ARRAY ? !Array.isArray(value) : (typeof value !== propDef.type)) {
+                throw new TypeError(`Prop is not of type ${ propDef.type }: ${ key }`);
+            }
+        } else {
+            if (propDef.required !== false) {
+                throw new Error(`Expected prop "${ key }" to be defined`);
+            }
+        }
+
         if (__DEBUG__ && isDefined(value) && propDef.validate) {
             // $FlowFixMe[incompatible-call]
             // $FlowFixMe[incompatible-exact]
-            propDef.validate({ value, props });
+            propDef.validate({ value, props: newProps });
         }
 
         if (isDefined(value) && propDef.decorate) {
-            // $FlowFixMe[incompatible-call]
-            const decoratedValue = propDef.decorate({ value, props, state, close, focus, event, onError });
-            // $FlowFixMe[incompatible-type]
-            props[key] = decoratedValue;
+            // Store the value the first time the decorate props is executed and use it in further executions
+            // This will be delete if the component instance is cloned
+            // $FlowFixMe
+            const originalValue = getOrSet(propDef, '_originalDefinition', () => value);
```

Comment:
Yeah this is going to be really problematic -- since the same `propDef` object could be shared between many different instances of a component that are instantiated/rendered at the same time.

Lemme think through if there's a better alternative, but if we want to store the original value I think we should find a scope that's local to the current component instance.

Reply:
FYI: added a few tests which cover multiple component instances, and multiple `updateProps` calls, please see https://github.com/krakenjs/zoid/commit/044028fcfb199707bf388715f55f1871aef0c23e

Reply:
OK, pushed up a new branch/commit which avoids the multi-decorate pattern and also passes all of these new tests. Please let me know if this looks ok to you: https://github.com/krakenjs/zoid/commit/c4bba5c1638cca17b618b99abf01cf3804520df4

Reply:
Looks like a cleaner code instead of storing all the time a function reference.